orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/ClientSessionService.java

```diff
@@ -94,22 +97,33 @@ public Optional<ClientSession> getClientSession(String clientSessionId) {
         }
     }
 
-    public void saveClientSession(String clientSessionId, ClientSession clientSession) {
+    public void updateStoredClientSession(String clientSessionId, ClientSession clientSession) {
         attachLogFieldToLogs(CLIENT_SESSION_ID, clientSessionId);
         attachLogFieldToLogs(GOVUK_SIGNIN_JOURNEY_ID, clientSessionId);
 
+        var clientSessionKey = CLIENT_SESSION_PREFIX.concat(clientSessionId);
+
+        if (!redisConnectionService.keyExists(clientSessionKey)) {
+            LOG.error(
+                    "Couldn't update client session with given key as it was not present in redis");
+            throw new IllegalArgumentException(
+                    format("Client session with ID {0} not found.", clientSessionId));
+        }
+
         try {
-            redisConnectionService.saveWithExpiry(
-                    CLIENT_SESSION_PREFIX.concat(clientSessionId),
-                    objectMapper.writeValueAsString(clientSession),
-                    configurationService.getSessionExpiry());
-        } catch (JsonException e) {
+            var oldClientSession = redisConnectionService.getValue(clientSessionKey);
+            var newClientSession = objectMapper.writeValueAsString(clientSession);
+            var updatedClientSession =
```

Comment:
This is going to have a big performance hit, and lead to some nasty race conditions, right?

Reply:
I don't know how bing the performance hit would be (we could add a local cache for the raw json string). Strictly speaking best practice would be to use a distributed lock, but I thought it was unnecessary. If we ever were somehow updating a client session in two places on a journey, the way we were updating is already problematic.